### ISSUE TITLE
for BERT pretraining, layer deeper than 6 is unable to converge in your default parameter

### ISSUE THREAD

**Issue by gaopengcuhk** [facebookresearch/XLM#159]
Hi, I run the following code
Command 1, 8 GPU training for 6 layer BERT
srun -p clusterRTX --gres=gpu:8 --ntasks=8 --ntasks-per-node=8 --cpus-per-task=3 python train.py --exp_name xlm_en --dump_path ./dumped   --data_path ./data/processed/XLM_en/30k/ --lgs 'en' --clm_steps '' --mlm_steps 'en' --emb_dim 2048  --n_layers 6 --n_heads 16 --dropout 0.1  --attention_dropout 0.1 --gelu_activation true --batch_size 8 --bptt 256 --optimizer adam,lr=0.0001  --epoch_size 300000 --max_epoch 100000 --validation_metrics _valid_en_mlm_ppl --stopping_criterion _valid_en_mlm_ppl,25 --fp16 false  --word_mask_keep_rand '0.8,0.1,0.1' --word_pred '0.15'

The loss will converge soon.

Command 2, 8 GPU training for 8 layer BERT
srun -p clusterRTX --gres=gpu:8 --ntasks=8 --ntasks-per-node=8 --cpus-per-task=3 python train.py --exp_name xlm_en --dump_path ./dumped   --data_path ./data/processed/XLM_en/30k/ --lgs 'en' --clm_steps '' --mlm_steps 'en' --emb_dim 2048  --n_layers 8 --n_heads 16 --dropout 0.1  --attention_dropout 0.1 --gelu_activation true --batch_size 8 --bptt 256 --optimizer adam,lr=0.0001  --epoch_size 300000 --max_epoch 100000 --validation_metrics _valid_en_mlm_ppl --stopping_criterion _valid_en_mlm_ppl,25 --fp16 false  --word_mask_keep_rand '0.8,0.1,0.1' --word_pred '0.15'

The loss will go around 7.5 even for days of training.

**Comment by glample** [comment]
Yes, with more layers training usually becomes more difficult. However, the problem usually start with 20 or 24 layers, for 8 layers I'm surprised that this does not work for you. Can you try with larger batches and see if it helps? Something like: `--accumulate_gradients 4` to make batches 4x larger. 

More importantly, linear warmup for the learning rate usually helps **a lot**. As well as weight decay. I would recommend something like this:
``--optimizer adam_inverse_sqrt,warmup_updates=20000,lr=0.0001,weight_decay=0.01``

**Comment by aconneau** [comment]
As @glample 
 pointed out, I would try also: optimizer=adam_inverse_sqrt,lr=0.00010,warmup_updates=30000,beta1=0.9,beta2=0.999,weight_decay=0.01,eps=0.000001

**Comment by gaopengcuhk** [comment]
Currently, I am running 8 layer bert pretraining with adam optimizer. I will test whether the provided suggestion can solve deeper model in the next week.
